### DISCUSSION TITLE
Webhook timesout even though code takes under 10s

### DISCUSSION THREAD

**Discussion by mah51** [vercel/vercel#6537] [Help]
Hi my vercel function (nextjs pages/api route) is return 504 gateway times out, even though the whole function takes around 170ms to execute (timed with console.time() .timeEnd()). 

Any idea why this is?

**Comment by leerob** [comment]
Hello! Can you reproduce this or providing logging showing the 170ms?

**Reply by mah51** [reply]
Hi Lee, thanks for the reply.

This is the error displayed in vercel's console:

```
[POST] /api/report
15:28:59:33
{  
    player: '74567547654', 
    reportingPlayer: '745674567',  
    reason: 'Cheating',  additional: '475674567',  
    reportingPlayerImage: 'https://cdn.discordapp.com/avatars/143699512790089729/f05dc4554e33e9ca5168d2ea2f25a3ee.png'
}
204
report: 162.183ms
2021-08-01T14:29:09.395Z a57a54e3-0432-4fa4-95e2-4a0b0438200d Task timed out after 10.01 seconds
```

and here is the code of the whole function:

```
const handler = async (req, res) => {
  console.time('report');

  const data = JSON.parse(req.body);
  console.log(data);
  const payload = {
    embeds: [
      {
        timeStamp: new Date(Date.now()).toISOString(),
        username: 'The bad boi exterminator',
        avatar_url:
          'https://cdn.discordapp.com/icons/866703918213759017/a8ccdce506e66a686fab20ec74a751b0.webp?size=128',
        title: data.player + ' reported for ' + data.reason,
        color: '12405557',
        description: data.additional || 'Do something about it mate',
        footer: {
          text: 'Reports provided by the best boi about',
        },
        author: {
          name: data.reportingPlayer,
          icon_url: data.reportingPlayerImage
            ? data.reportingPlayerImage
            : null,
        },
      },
    ],
  };

  const response = await fetch(process.env.DISCORD_WEBHOOK_URL, {
    method: 'POST',
    headers: {
      'Content-Type': 'application/json',
    },
    body: JSON.stringify(payload),
  });
  console.log(response.status);
  res.status(response.status).json({ message: 'success' });
  return console.timeEnd('report');
};

export default handler;
```

At the base of it, its just a simple post request so im not sure why its taking so long.

Thanks for the help in advance :).

**Reply by leerob** [reply]
Can you return res.status instead of the console time?

**Reply by mah51** [reply]
Yeah,  i swapped them and the same error occured 

```
[POST] /api/report
16:32:38:40
{ 
 player: '542354234523', 
 reportingPlayer: '523452354', 
 reason: 'Cheating',  
additional: '23523452354', 
 reportingPlayerImage: 'https://cdn.discordapp.com/avatars/143699512790089729/f05dc4554e33e9ca5168d2ea2f25a3ee.png'
}
204
report: 190.458ms2021-08-01T15:32:48.471Z 73cb1e2b-6534-4efa-92ca-eab6d754d47a Task timed out after 10.01 seconds
```

**Reply by leerob** [reply]
Can you post the vercel project / github link?

**Reply by mah51** [reply]
Yeah of course:
https://rustybois.com is the website
https://github.com/mah51/rusty-bois-web is the repo

The site is for a rust game server, and the idea of the this rest function is to allow players to be reported. The button on the nav opens a modal which once filled out sends an api request to this endpoint. The ReportPlayerButton component is in Nav.js. Apologies about the code being messy, Ive been rushing to get it done! Thanks :)

**Reply by leerob** [reply]
Looks like the repo is private. Could you open it up?

**Reply by mah51** [reply]
Should be public now, my bad!

**Reply by leerob** [reply]
Everything looks okay, I'm not sure what is happening here. Can you try slowing commenting out code or adding more logging? Does it work as expected locally? Can you confirm the webhook URL is properly set?

**Reply by mah51** [reply]
I know the webhook URL is correctly set as it sometimes logs to the channel occasionally and i have tried console.logging it. It works as expected locally (almost instantly) however apart from commenting out the post request i cant really think of another way to highlight the issue. But i will do so now.

**Reply by mah51** [reply]
Oh wow, that is interesting...

I commented out the fetch() call and it still has the same issue:

```
const handler = async (req, res) => {
  console.time('report');

  const data = JSON.parse(req.body);
  console.log(data);
  const payload = {
    embeds: [
      {
        timeStamp: new Date(Date.now()).toISOString(),
        username: 'The bad boi exterminator',
        avatar_url:
          'https://cdn.discordapp.com/icons/866703918213759017/a8ccdce506e66a686fab20ec74a751b0.webp?size=128',
        title: data.player + ' reported for ' + data.reason,
        color: '12405557',
        description: data.additional || 'Do something about it you fat fucks',
        footer: {
          text: 'Reports provided by the best boi about',
        },
        author: {
          name: data.reportingPlayer,
          icon_url: data.reportingPlayerImage
            ? data.reportingPlayerImage
            : null,
        },
      },
    ],
  };

  // const response = await fetch(process.env.DISCORD_WEBHOOK_URL, {
  //   method: 'POST',
  //   headers: {
  //     'Content-Type': 'application/json',
  //   },
  //   body: JSON.stringify(payload),
  // });

  const response = {
    status: 204,
    payload,
  };
  console.log(response.status);
  console.timeEnd('report');
  return res.status(response.status).json({ message: 'success' });
};

export default handler;
```

Will try a few more bits

**Reply by mah51** [reply]
Ok I have found the problem. Without any logic in the function, if i return res.status(204).json(...) the function times out. However if i return res.status(200).json(...) it works as expected. I have no idea why this is the case (it was return 204 as discord returns that from their api). 

I am now using this logic instead:

```
  if (response.status < 300) {
    return res.status(200).json({ message: 'success' });
  }

  return res.status(500).json({ message: 'error' });
```

Thank you very much for the help Lee, I love the work you do for this community! (especially the tutorials :D)

**Reply by leerob** [reply]
Confirmed this is a bug on our side, will get it fixed! This workaround is great too 😄 Guessing it's because 204 is expecting no content, but you're also forwarding JSON.

**Reply by mah51** [reply]
Yeah, that makes sense. Thank you!
